docs/source_rewriter.md

```diff
@@ -0,0 +1,210 @@
+# Source Rewriter Design
+
+In Phase I we used a custom compiler to insert call gates at compartment
+transitions. In Phase II we treat dynamic shared objects (DSOs) as compartments
+and use standard linking techniques to insert call gates between
+intercompartment direct calls. For indirect calls we relied on user annotations
+on function pointers. Instead of adding function pointer annotations we instead
+plan on creating a tool that rewrites source files to make the necessary
+changes.
+
+## Rationale
+
+Our initial approach relied on analyzing and rewriting header files to (1)
+create an output `.c` for the call gate wrappers and (2) assist the user in
+adding the function pointer annotations (by rewriting function pointer types to
+trigger compiler errors where changes are needed). This approach has two issues.
+
+First a header's AST can change depending on the preprocessor context when
+it's included. We currently create one rewritten `.h` for each input `.h`
+which can have correctness issues in some rare cases. While we can fix the
+correctness issue by creating one rewritten `.h` for each `#include` this
+engineering problem will quickly become messy. We also use the same set of `-D`
+for all `.h`s, but the correct solution is to use the exact set of flags used to
+compile the `.c` which includes each `.h`.
+
+The second problem is that code that makes heavy use of function pointers is
+burdensome to annotate and they degrade readability. Even in the case where only a
+small set of the function pointers in a code base are used, the rewritten `.h`s
+trigger compiler errors at every site without annotations. Also some of the
+annotations need the pointer's function signature specified with the mangled
+type name which is a non-starter.
+
+## Goals
+
+By analyzing and rewriting `.c`s we can avoid these issues. The goal here will
+be to (1) create an output `.c` for call gate wrappers and (2) insert simplified
+function pointer annotations in the input `.c`s. We'll initially focus on the
+case with one rewritten `.c` for each input `.c` and attempt to insert the
+annotations such that the rewritten `.c` can be compiled with the same flags.
+
+## Source rewriting process
+
+The goals above can be broken down into these subgoals.
+
+1. Define call gates for direct compartment transitions
+2. Add call gates for pointers
+3. Add call gates for indirect callsites
+
+The rewriter will take all source files as input along with which compartment
+each belongs to. It will generate one output `.c` and one `.h` for use with all
+compartments. It will also generate compartment-specific linker args and
+`objcopy --redefine-syms` files.
+
+### Define call gates for direct transitions
+
+The rewriter will work in two phases. In the first phase it'll parse all source
+files and keep track of functions defined and their compartment, functions
+called and from which compartment (TODO: Should we check functions called or
```

Comment:
If we create call gates for undefined functions in each translation unit, then we will almost certainly have more call gates than we actually need but we'll only need to rerun the rewriter when more functions are declared (e.g. a new #include is added).

If we create call gates for functions that are called, then we'll have the minimum number of call gates we need, but we'll need to rerun the rewriter each time a new function call to another compartment is added.

Reply:
Went with the first option in #182 for now.